### PR TITLE
python-tabulate: migrate to Python@3.10

### PR DESCRIPTION
**Pull request by iMichka**
This aligns python-tabulate with other formulae like six, pillow and numpy,
which are "multi-python" to ease migrations.
Python-tabulate is used by multiple formulae and this will help decoupe
some part of the dependency tree, thus easing the migration.

- [ ] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [ ] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [ ] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

### CODE REVIEW

**Review by carlocab — COMMENTED** [review]
The `tabulate` CLI tool definitely needs a Python at runtime.

**Comment by iMichka** [comment]
Fix for anime-downloaded: #99375

File: Formula/libpython-tabulate.rb
Comment (cho-m) [inline Formula/libpython-tabulate.rb:11]: This won't work cleanly. It is the linking that allows dependents to find Python package due to default site-packages.

If you want to keep keg-only, you will need to create .pth files for every dependent's venv so it can find `tabulate`.

The easier option is to follow `six` as an example and link formula.

File: Formula/python-tabulate.rb
Comment (cho-m) [inline Formula/python-tabulate.rb:25]: Alternatively, remove `keg_only` from `libpython-tabulate`, add it as a dependency, and then just install the Python script via something like:
```suggestion
   system "python3", "setup.py", "--no-user-cfg", "install_scripts", "--install-dir=#{bin}", "--skip-build"
```

Or some other variation based on `Language::Python.setup_install_args`.

This will only install the script using Python@3.10. The functionality will depend on library linked in `libpython-tabulate`.

File: Formula/libpython-tabulate.rb
Comment (cho-m) [inline Formula/libpython-tabulate.rb:47]: These aren't used
```suggestion
```

**Comment by iMichka** [comment]
Needs #102092

File: Formula/libpython-tabulate.rb
Comment (iMichka) [reply Formula/libpython-tabulate.rb]: Done

File: Formula/python-tabulate.rb
Comment (iMichka) [reply Formula/python-tabulate.rb]: Done

**Comment by cho-m** [comment]
Some revision bumps were lost in rebase.

File: Formula/klee.rb
Comment (carlocab) [inline Formula/klee.rb:19]: I don't think this is right. This formula definitely uses `tabulate` by shelling out to it.

**Review by carlocab — COMMENTED** [review]
This seems wrong for at least one formula; maybe more.

Also not a fan of the duplication in code and possible duplication in installs on users' systems.

File: Formula/klee.rb
Comment (iMichka) [reply Formula/klee.rb:19]: Can you be more precise, I could not find where this is happening? I only found the `tools/klee-stats/klee-stats` file that imports the library.

**Comment by iMichka** [comment]
> Some revision bumps were lost in rebase.

Done

> This seems wrong for at least one formula; maybe more.

I checked a few and they all seem to only import the library. We can still change the dependency back for any formula if a user complains.

> Also not a fan of the duplication in code and possible duplication in installs on users' systems.

Yeah, agreed. But I see no better way.
I think all of these were vendored initially, so we removed some duplication. I'd rather have 2 formulae for this than 10 vendor/resource blocks.

But we made our life more complicated for Python migrations by removing the vendoring (the same holds for `six` and so on). And it's not going to be better in the long term, so simplifying the dependency tree looks tempting to me.

**Comment by iMichka** [comment]
Needs #102950

**Comment by iMichka** [comment]
`klee` fails with
> error: invalid application of 'sizeof' to an incomplete type 'struct stat64'

@carlocab @danielnachun there was a similar error in #79467, do you remember what change fixed it?

**Comment by cho-m** [comment]
`klee` needs some `rewrite_shebang`.

`cryptominisat` test failed on ARM
```
==>/opt/homebrew/opt/python@3.10/bin/python3 test.py
  Traceback (most recent call last):
    File "/private/tmp/cryptominisat-test-20220608-74227-vpyvly/test.py", line 1, in <module>
      import pycryptosat
  ModuleNotFoundError: No module named 'pycryptosat'
```

**Review by cho-m — DISMISSED** [review]
I'm okay with this.

Not sure what `tabulate` script usage @carlocab was referring to, so may want to finalize on that before merge.

**Review by SMillerDev — DISMISSED** [review]

**Comment by iMichka** [comment]
`regepy` got updated meanwhile ... let's re-run this ...

**Comment by cho-m** [comment]
@iMichka, this needs another rebase due to `esphome` #103828

**Comment by iMichka** [comment]
Rebased. I hope we can merge this as soon as the build is green to avoid any further delay.

Regarding `klee`, we can still swith to the `tabulate` formula that provides the binary in case I missed something, but it looks like everything is fine with the current setup.

**Review by cho-m — DISMISSED** [review]

**Comment by danielnachun** [comment]
Unfortunately I think this needs the self-hosted runner as it timed out on Linux.

**Comment by danielnachun** [comment]
I restarted this again.  We should publish the bottles as soon as the run is done, so any maintainers should feel free to do so if I don't get to it first.

**Comment by cho-m** [comment]
Another conflict due to `localstack` merge #103936

**Comment by iMichka** [comment]
Rebased again ...

**Review by danielnachun — DISMISSED** [review]
Reapproving again so we can merge this ASAP.

**Review by cho-m — DISMISSED** [review]

**Comment by cho-m** [comment]
Another `esphome` update #103970. May just want to disable automerge workflow for half a day as we need ~10 hours to get the CI run and merge.

**Comment by danielnachun** [comment]
I decided to restart https://github.com/Homebrew/homebrew-core/pull/97618 because everything there is ready to go and it is blocking several PRs.  However that means we do need to wait to rerun this.  As @cho-m said I think we can't realistically merge this without temporarily disabling automerge so that something doesn't get updated during the CI run.

**Review by SMillerDev — DISMISSED** [review]

**Comment by iMichka** [comment]
Let's try once more.

Even if we disable automerge, it does not prevent a maintainer to merge manually anything that might break this PR.
I will notify the maintainers on slack to watch out for any conflicts for the next hours.

**Comment by cho-m** [comment]
That's fine.

If we end up hitting another merge conflict, we may want to break up PR into smaller changes that might not need long timeout, e.g.
1. `libpython-tabulate` + `python-tabulate` on Python 3.9
2. migrate dependents to `libpython-tabulate` in separate PRs.
3. migrate `python-tabulate` to Python 3.10

This does cause a double revision bump on `python-tabulate` so would be better to get this merged in single PR if possible.

**Comment by iMichka** [comment]
I think the latest `setuptools_scm` updated on pypi 2 hours ago broke everything ...

**Comment by cho-m** [comment]
Yeah, I hit a similar issue in `jupyterlab` PR. There is a build loop when only using source due to `setuptools_scm -> importlib_metadata -> setuptools_scm[toml]`.

**Comment by danielnachun** [comment]
> That's fine. We'll probably want to wait on this before running `llvm` PR then.
> 
> If we end up hitting another merge conflict, we may want to break up PR into smaller changes that might not need long timeout, e.g.
> 
> 1. `libpython-tabulate` + `python-tabulate` on Python 3.9
> 2. migrate dependents to `libpython-tabulate` in separate PRs.
> 3. migrate `python-tabulate` to Python 3.10
> 
> This does cause a double revision bump on `python-tabulate` so would be better to get this merged in single PR if possible.

I think this is a much easier way of getting this merged and is worth the double revision bump.  It will take a while for us to actually migrate all the individual dependencies anyway, so the two revisions will be spaced out by at least a couple of days.  We should update and migrate `klee` first so that it doesn't block the LLVM PR.

**Comment by cho-m** [comment]
Cancelling run as all runners are going to hit `setuptools_scm` issue. Looks like there is now an open upstream issue https://github.com/pypa/setuptools_scm/issues/722

**Comment by cho-m** [comment]
ARM Monterey runner didn't run correctly.

`dvc` failed on ARM Big Sur due to `setuptools_scm_git_archive`:
```
Collecting setuptools_scm_git_archive==1.1
      Downloading setuptools_scm_git_archive-1.1.tar.gz (3.0 kB)
      Preparing metadata (setup.py): started
      Preparing metadata (setup.py): finished with status 'done'
      WARNING: Requested setuptools_scm_git_archive==1.1 from https://files.pythonhosted.org/packages/7e/2c/0c15b29a1b5940250bfdc4a4f53272e35cd7cf8a34159291b6b4ec9eb291/setuptools_scm_git_archive-1.1.tar.gz#sha256=6026f61089b73fa1b5ee737e95314f41cb512609b393530385ed281d0b46c062, but installing version 0.0
    Discarding https://files.pythonhosted.org/packages/7e/2c/0c15b29a1b5940250bfdc4a4f53272e35cd7cf8a34159291b6b4ec9eb291/setuptools_scm_git_archive-1.1.tar.gz#sha256=6026f61089b73fa1b5ee737e95314f41cb512609b393530385ed281d0b46c062 (from https://pypi.org/simple/setuptools-scm-git-archive/): Requested setuptools_scm_git_archive==1.1 from https://files.pythonhosted.org/packages/7e/2c/0c15b29a1b5940250bfdc4a4f53272e35cd7cf8a34159291b6b4ec9eb291/setuptools_scm_git_archive-1.1.tar.gz#sha256=6026f61089b73fa1b5ee737e95314f41cb512609b393530385ed281d0b46c062 has inconsistent version: filename has '1.1', but metadata has '0.0'
    ERROR: Could not find a version that satisfies the requirement setuptools_scm_git_archive==1.1 (from versions: 1.0, 1.1, 1.2)
    ERROR: No matching distribution found for setuptools_scm_git_archive==1.1
    WARNING: There was an error checking the latest version of pip.
    error: subprocess-exited-with-error
```

**Comment by iMichka** [comment]
I'm going for the double revision bump / 2 step approach, else it will be hard to get this shipped.

**Comment by branchvincent** [comment]
We have only 4 formulae left migrate to `libpython-tabulate`:

* `anime-downloader`: #104338
* `aws-google-auth`: #104340
*  `dvc`: #104672
* `klee`: #104288

**Comment by danielnachun** [comment]
Let's use the allowlist to get those 4 merged.  We'll end up doing a cleanup of the conflicting versions allowlist anyway after splitting out `glib` and `glib-utils`.

File: Formula/klee.rb
Comment (carlocab) [inline Formula/klee.rb:24]: Does this need `llvm@13`?

File: Formula/klee.rb
Comment (carlocab) [inline Formula/klee.rb:42]: This might not work -- I think upstream LLVM nixed the ability to do standalone builds of `libc++`.

**Review by carlocab — APPROVED** [review]
Fine with the approach here, but there are conflicts that need resolving. 😢

**Comment by iMichka** [comment]
Closing as only `klee` remains to be migrated: #104288

**Pull request closed without merging**
